File: lib/scraping/url-rules.ts
import type { SourceRule } from "@/lib/types";

export function normaliseUrl(href: string, baseUrl: string) {
  try {
    const url = new URL(href, baseUrl);
    url.hash = "";
    return url.toString();
  } catch {
    return null;
  }
}

export function matchesRule(url: string, rule: SourceRule) {
  try {
    return new RegExp(rule.pattern).test(url);
  } catch {
    return url.includes(rule.pattern);
  }
}

export function shouldVisitUrl(url: string, rules: SourceRule[]) {
  const activeRules = rules.filter((rule) => rule.is_active);
  const excludes = activeRules.filter((rule) => rule.rule_type === "exclude");
  const includes = activeRules.filter((rule) => rule.rule_type === "include");

  if (excludes.some((rule) => matchesRule(url, rule))) {
    return false;
  }

  if (includes.length === 0) {
    return true;
  }

  return includes.some((rule) => matchesRule(url, rule));
}

export function classifyUrl(url: string, rules: SourceRule[]) {
  const activeRules = rules.filter((rule) => rule.is_active);

  if (activeRules.some((rule) => rule.rule_type === "detail" && matchesRule(url, rule))) {
    return "detail" as const;
  }

  if (activeRules.some((rule) => rule.rule_type === "listing" && matchesRule(url, rule))) {
    return "listing" as const;
  }

  if (/award|awarded|contract/i.test(url)) {
    return "award" as const;
  }

  if (/\.pdf($|\?)/i.test(url)) {
    return "document" as const;
  }

  return "unknown" as const;
}

export function matchedRuleLabel(url: string, rules: SourceRule[]) {
  const rule = rules
    .filter((candidate) => candidate.is_active)
    .sort((a, b) => a.priority - b.priority)
    .find((candidate) => matchesRule(url, candidate));

  return rule ? `${rule.rule_type}:${rule.pattern}` : null;
}

File: worker/app/pipeline/crawl/url_rules.py
import re
from urllib.parse import urldefrag, urljoin


def normalise_url(href: str, base_url: str) -> str | None:
    try:
        joined = urljoin(base_url, href)
        return urldefrag(joined).url
    except ValueError:
        return None


def matches_rule(url: str, rule: dict) -> bool:
    pattern = rule.get("pattern", "")
    try:
        return re.search(pattern, url) is not None
    except re.error:
        return pattern in url


def should_visit_url(url: str, rules: list[dict]) -> bool:
    active = [rule for rule in rules if rule.get("is_active", True)]
    excludes = [rule for rule in active if rule.get("rule_type") == "exclude"]
    includes = [rule for rule in active if rule.get("rule_type") == "include"]

    if any(matches_rule(url, rule) for rule in excludes):
        return False

    if not includes:
        return True

    return any(matches_rule(url, rule) for rule in includes)


def classify_url(url: str, rules: list[dict]) -> str:
    active = [rule for rule in rules if rule.get("is_active", True)]

    if any(rule.get("rule_type") == "detail" and matches_rule(url, rule) for rule in active):
        return "detail"

    if any(rule.get("rule_type") == "listing" and matches_rule(url, rule) for rule in active):
        return "listing"

    if re.search(r"award|awarded|contract", url, re.IGNORECASE):
        return "award"

    if re.search(r"\.pdf($|\?)", url, re.IGNORECASE):
        return "document"

    return "unknown"


def matched_rule_label(url: str, rules: list[dict]) -> str | None:
    active = sorted(
        [rule for rule in rules if rule.get("is_active", True)],
        key=lambda rule: rule.get("priority", 100),
    )

    for rule in active:
        if matches_rule(url, rule):
            return f"{rule.get('rule_type')}:{rule.get('pattern')}"

    return None
